source/nvwave.py

```diff
@@ -223,14 +331,17 @@ def _feedUnbuffered(self, data, onDone=None):
 		with self._lock:
 			with self._waveout_lock:
 				self.open()
-				with self._global_waveout_lock:
-					winmm.waveOutPrepareHeader(self._waveout, LPWAVEHDR(whdr), sizeof(WAVEHDR))
 				try:
 					with self._global_waveout_lock:
+						winmm.waveOutPrepareHeader(self._waveout, LPWAVEHDR(whdr), sizeof(WAVEHDR))
 						winmm.waveOutWrite(self._waveout, LPWAVEHDR(whdr), sizeof(WAVEHDR))
-				except WindowsError as e:
-					self.close()
-					raise e
+				except WindowsError:
+					log.info("Error during feed. Resetting the device.")
+					self._close()  # don't try to call stop on a "broken" device.
+					self._setCurrentDevice(self._preferredDeviceName)
+					self.open()
+					self.feed(data, onDone)

```

Comment:
what would happen if feed / _feedUnbuffered was to always raise WindowsError... say with a buggy driver. This would cause a never ending recursive exception loop. Obviously NVDA could never recover audio at this point, but  we don't however want to end up in that kind of loop...

Reply:
It's possible, but for that to happen, opening the device would have to succeed, but writing to it would have to fail. It seems unlikely, but it is easy to protect against. I'll do this.